### ISSUE TITLE
`zfs_read` uses `mappedread` for files that are mapped read-only

### ISSUE THREAD

**Issue by dominikh** [openzfs/zfs#16031]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| NixOS
Distribution Version	| 
Kernel Version	| Linux 6.6.16
Architecture	| amd64
OpenZFS Version	| zfs-2.2.2-1, zfs-kmod-2.2.2-1

### Describe the problem you're observing

`zfs_read` calls `mappedread` for reads from files that are memory mapped: https://github.com/openzfs/zfs/blob/8cd8ccca5383dcdd9bf55d4d22921a6b43b4ebe1/module/zfs/zfs_vnops.c#L305-L311

This happens even if all mappings are read-only and there can be no dirty cached pages. This is problematic because `mappedread` is considerably slower than the fast path. On my system, a 32 MB read of cached data completes in ~4.2ms on the fast path and in ~19.5ms on the slow path, a 4.4x slowdown.

### Describe how to reproduce the problem

`read.c`:
```c
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <time.h>
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>

int main(int argc, char *argv[]) {
  int fd = open(argv[1], O_RDONLY);
  char *buf = malloc(32*1000*1000);
  while (true) {
	struct timespec start, end;
	clock_gettime(CLOCK_MONOTONIC, &start);

	ssize_t sz = pread(fd, buf, 32*1000*1000, 0);

	clock_gettime(CLOCK_MONOTONIC, &end);
	uint64_t delta_us = (end.tv_sec - start.tv_sec) * 1000000 + (end.tv_nsec - start.tv_nsec) / 1000;
	printf("%ld %ld\n", sz, delta_us);
  };
  return 0;
}
```

`map.c`:
```c
#include <stdio.h>
#include <sys/mman.h>
#include <unistd.h>
#include <fcntl.h>

int main(int argc, char *argv[]) {
  int fd = open(argv[1], O_RDONLY);
  void *addr = mmap(NULL, 32*1000*1000, PROT_READ, MAP_SHARED, fd, 0);
  printf("%p\n", addr);
  sleep(99999);
}
```

preparation:
```sh
dd if=/dev/urandom of=32meg bs=32M count=1
chmod ugo=r 32meg
gcc read.c -o read
gcc map.c -o map
```

Terminal 1:
```
$ ./read 32meg
32000000 13903
32000000 3992
32000000 3960
32000000 4160
32000000 4019
...
32000000 13735
32000000 19122
32000000 18995
32000000 19508
```

Terminal 2:
```
$ sleep 1; ./map 32meg
```

The file is both read-only and mapped with PROT_READ. In production, the file was further on a read-only mount.

I also find it worth noting that even after killing `map`, the reads will not speed up again, as if the "mappedness" of the file were sticky. This is true even across multiple runs of `read`. Once the file has been mapped once, reads will be slow. I haven't dug into why that's the case. Let me know if that's worth splitting into its own issue, or if a nuance of mappings is lost on me.

`zn_has_cached_data` is defined as https://github.com/openzfs/zfs/blob/8cd8ccca5383dcdd9bf55d4d22921a6b43b4ebe1/include/os/linux/zfs/sys/zfs_znode_impl.h#L80-L86

 Based on my kernel's version, I assume this is using the `filemap_range_has_page` path. In either case, the function behaves more like _may_ have cached data, not _has_ (this holds true for the FreeBSD implementation, too.)

**Comment by rrevans** [comment]
I'm able to easily reproduce this as well.

Linux tries to keep as many filesystem cached pages around as possible. `munmap` (or the equivalent at process termination) simply releases the mapping for that process, but the previously mapped pages remain in the pagecache and  associated with the file until they are reclaimed.

Because of this another process can remap the pages at any time possibly for writing (at least in the general case). ZFS does not have any special tracking for any of this, so it must read (and write!) any cached pages to keep them in sync with normal file IO operations.

That said, the performance being so much lower is somewhat surprising and maybe there are low-hanging optimizations?

`madvise` is a possible workaround to release pages sooner though its guarantees are somewhat weak.

**Comment by dominikh** [comment]
> Linux tries to keep as many filesystem cached pages around as possible. munmap (or the equivalent at process termination) simply releases the mapping for that process, but the previously mapped pages remain in the pagecache and associated with the file until they are reclaimed.

Is this specific to ZFS and the difference between the ARC and the page cache? It was my understanding that for other file systems, the page cache is used for file caching and mmaped files alike.

I've also tried `echo 3 > /proc/sys/vm/drop_caches` and this didn't improve the read speed, either, even though the `read` reproducer was the only program accessing the file, so I would have expected it to be dropped from the page cache after using drop_caches. Unfortunately I don't know of a fine-grained way of determining if any of the file is in the page cache. 

> That said, the performance being so much lower is somewhat surprising and maybe there are low-hanging optimizations?

Agreed. In fact, 32 MB over 19.5ms translates to 1.6 GB/s, which is even slower than the underlying storage.

**Comment by rrevans** [comment]
> Is this specific to ZFS and the difference between the ARC and the page cache? It was my understanding that for other file systems, the page cache is used for file caching and mmaped files alike.

Yes. ZFS does not use the page cache except for mmap support. And even then the page cache is a completely separate copy. Memory-mapped files will appear twice in memory, and ZFS incurs overhead to keep them in sync.

**Comment by rrevans** [comment]
@dominikh Looking at this closely, are you sure your ZFS build is picking up `filemap_range_has_page`?

While trying to look a bit closer, I found that my `zfs_config.h` did not have `HAVE_FILEMAP_RANGE_HAS_PAGE` defined.

Turns out the config test has been broken since v5.18 when `filemap_range_has_page` moved to `<linux/pagemap.h>` and the test only includes `fs.h`:

https://github.com/openzfs/zfs/blob/e39e20b6dc73bd7df1f097c23b5297bcc989ed53/config/kernel-filemap.m4#L1-L9

**TL;DR** since Linux 5.18 `zn_has_cached_data` has been falling back to the `z_is_mapped` version because of symbol churn.

This explains the extra-sticky behavior. The "was ever mapped" bit will stay set until the inode cache drops the file entry. Probably explains why `drop_caches` didn't quite work as expected.

**Comment by rrevans** [comment]
I created a little proof of concept patchset: rrevans@df6800fda and ran some tests.

**TL;DR** #16034 by itself helps, and there are performance gains to be had for mapped files if pagecache routines were to be used directly or the fallback path were optimized.

Test method:
1) Write a new 1 GiB random file
2) `zpool sync` and `sysctl vm.drop_caches=3`
2) Read the file 3 times with `dd bs=128k` and take median result
3) mmap the file (and maybe actually read the pages, see below)
4) Read file again 3 times with `dd` and take median speed

At head, just mmaping the file _and not touching any pages_ same as the original report:
* Before mmap: 6.9 GB/s
* After mmap: 1.8 GB/s

At head, mmaping the file but touching all pages first during mmap step:
* Before mmap: 7.3 GB/s
* After mmap: 10.6 GB/s

So, it seems `mappedread` is much more expensive for pages ZFS thinks are mapped but no pages are actually cached. This is probably because it falls back to individual page-sized reads within `dmu_read_uio_dbuf`.

After #16034 only:

mmap only, no cached pages:
* Before mmap: 6.7 GB/s
* After mmap: 6.8 GB/s

mmap with cached pages:
* Before mmap: 6.7 GB/s
* After mmap: 10.7 GB/s

(But #16034 might not help much if the mmap'ed file has a large mix of cached and uncached pages.)

With my experimental patch:

mmap only, no cached pages:
* Before mmap: 7.1 GB/s
* After mmap: 7.0 GB/s

mmap with cached pages:
* Before mmap: 6.9 GB/s
* After mmap: 13.4 GB/s

**Comment by malventano** [comment]
I just tripped over this where a vllm docker container loading a large model from mapped storage on a local nvme ssd zpool was limited to ~600 MB/s, where that storage can otherwise go ~10x that for the same workload.
